src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -1339,6 +1343,8 @@ void EmitBackreference(RegexNode node)
             // Emits the code for an if(backreference)-then-else conditional.
             void EmitBackreferenceConditional(RegexNode node)
             {
+                Debug.Assert(node.Type is RegexNode.Testref, $"Unexpected type: {node.Type}");
```

Comment:
is it worth asserting child count whereever asserting node type ? eg you do for EmitAlternation above.